### ISSUE TITLE
lab04: neural networks

### ISSUE THREAD

**Issue by Revilllo7** [Revilllo7/computational-intelligence-class#44]
Do the following tasks:

info: preferably put the algorithms in a standalone directory and automate some functionality for less code repeating:

```
algorithms/
functions/(?)
lab01
lab02
lab03
...etc
```

## task01: neural networks from the theory side
simulate the behaviour and learning of a neural network following the `neural_network_backpropagation_pl.pdf`

- forward propagation
- check results
- add learning function: backward error propagation, gradient decrease, MSE (mean squared error), learning speed eta = 0.1
- check how the weights will be adjusted after the learning algorithm. Test if they match the script values

**!!! Will be on a test !!!**

## task02: pytorch
- utilise `DataLoading` edition and normalisation
- build a NL with a "sensible" structure. What structure is "sensible"? Explain in a comment, topology, activation function, etc.
- create training loop in PyTorch
- visualise the learning process on the learning curve. Two plots: `loss` (validation and training) and `accuracy` (validation and training)
- show accuracy and confusion matrix for the entire validation dataset
- interpret the results

---

## Note:
3 sets:
- training (to train)
- testing (at the end)
- validation (after training)

Visualise accuracy results across epochs for training and validation set
- y-axis accuracy
- x-axis epochs

Visualise loss results across epochs for training and validation set
- y-axis loss
- x-axis epochs

**Compare the training and validation plots, cause that gives you a lot of insight into how the Neural Network is**
over-trained? under-trained? just-right-trained?

Test different loss functions.
II Ii
II I_

**DO NOT OVERTRAIN**

### Create a function that creates a Neural Network
(adjustable for parameters) and just use that for task02 and task03. WAY better code and maintainability

---

Avoid unstable training. Worse accuracy, but stable learning is better!

## task03: diagnosis with a neural network
repeat the experiment from task02 for `diagnosis.csv`
